extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elastic/impl/ElasticProcessor.java

```diff
@@ -0,0 +1,232 @@
+        if (traverser instanceof ElasticProcessor.ElasticScrollTraverser) {
+            ElasticScrollTraverser<T> scrollTraverser = (ElasticScrollTraverser<T>) traverser;
+            scrollTraverser.close();
+        }
+    }
+
+    static class ElasticScrollTraverser<R> implements Traverser<R> {
+
+        private final ILogger logger;
+
+        private final RestHighLevelClient client;
+        private final FunctionEx<? super ActionRequest, RequestOptions> optionsFn;
+        private final FunctionEx<? super SearchHit, R> mapHitFn;
+        private final String scrollKeepAlive;
+        private final ConsumerEx<? super RestHighLevelClient> destroyFn;
+
+        private SearchHits hits;
+        private int nextHit;
+        private String scrollId;
+
+        ElasticScrollTraverser(ElasticSourceBuilder<R> builder, RestHighLevelClient client, SearchRequest sr,
+                               ILogger logger) {
+            this.client = client;
+            this.optionsFn = builder.optionsFn();
+            this.mapHitFn = builder.mapHitFn();
+            this.destroyFn = builder.destroyFn();
+            this.scrollKeepAlive = builder.scrollKeepAlive();
+            this.logger = logger;
+
+            try {
+                RequestOptions options = optionsFn.apply(sr);
+                SearchResponse response = this.client.search(sr, options);
+
+                // These should be always present, even when there are no results
+                hits = requireNonNull(response.getHits(), "null hits in the response");
+                scrollId = requireNonNull(response.getScrollId(), "null scrollId in the response");
+
+                TotalHits totalHits = hits.getTotalHits();
+                logger.fine("Initialized scroll with scrollId " + scrollId + ", total results " +
+                        totalHits.relation + ", " + totalHits.value);
+            } catch (IOException e) {
+                throw new JetException("Could not execute SearchRequest to Elastic", e);
+            }
+        }
+
+        @Override
+        public R next() {
+            if (hits.getHits().length == 0) {
+                scrollId = null;
+                return null;
+            }
+
+            if (nextHit >= hits.getHits().length) {
+                try {
+                    SearchScrollRequest ssr = new SearchScrollRequest(scrollId);
+                    ssr.scroll(scrollKeepAlive);
+
+                    SearchResponse searchResponse = client.scroll(ssr, optionsFn.apply(ssr));
+                    hits = searchResponse.getHits();
+                    if (hits.getHits().length == 0) {
+                        return null;
+                    }
+                    nextHit = 0;
+                } catch (IOException e) {
+                    throw new JetException("Could not execute SearchScrollRequest to Elastic", e);
+                }
+            }
+
+            return mapHitFn.apply(hits.getAt(nextHit++));
```

Comment:
you can use `Traverser.map()` where this traverser is called.

Reply:
Done.